decisions/0003-package-for-shared-server-features.md

```diff
@@ -0,0 +1,113 @@
+- Adapter (`@remix-run/{arc,cloudflare-pages,cloudflare-workers,express,fly,netlify,vercel}`)
+- Renderer (`@remix-run/react`)
+
+User code currently looks like this:
+```tsx
+// server runtime
+import { json, createFileSessionStorage } from "@remix-run/node";
+// adapter
+import { createRequestHandler } from "@remix-run/express"
+// renderer
+import { useLoaderData } from "@remix-run/react";
+```
+
+### Goals
+
+- make DX and docs simpler: more intuitive, less magic, eliminate some user errors
+- reduce bugs from `remix` internal complexity
+
+### Non-goals
+
+- maintainability of `remix` repo
+- more portable `app/` directory
+- reduce or increase number of packages
+
+Some of these are nice-to-haves, but not the primary reasons for this proposal.
+
+### Issues
+
+1. Users can import features from their runtime that aren't supported by their adapter:
+
+```tsx
+import { createFileSessionStorage } from "@remix-run/node";
+// Implementation only depends on server runtime, but _availability_ depends on the adapter!
+// ❌ This will crash in netlify nor in vercel!
+```
+
+2. Docs, examples, and videos refer to `@remix-run/node` by default for shared functionality like `json`:
+
+```tsx
+import { json } from "@remix-run/node";
+// Need docs to tell users that they should import from their corresponding runtime package, 
+// and not from `@remix-run/server-runtime`.
+```
+
+3. Keeping shared exports consistent across runtime packages is [error-prone](https://github.com/remix-run/remix/pull/3215) and tedious.
+
+## Decision
+
+Users should import features from _adapter packages_, not from runtime packages. Each adapter can export only the features it supports. Shared features will be imported from a new package, `@remix-run/server`.
+
+This means there won't be any imports needed from the runtime packages (`@remix-run/{cloudflare,deno,node`}) and these will become an implementation detail for most users. Only developers that need to develop adapters will import from the runtime packages directly.
+
+### User-facing packages
+
+Users still need to import from 4 Remix packages:
+
+- Dev `@remix-run/dev` (unchanged)
+- Shared server features `@remix-run/server`
+- Adapter-specific server features (`@remix-run/{arc,cloudflare-pages,cloudflare-workers,express,fly,netlify,vercel}`)
+- Renderer `@remix-run/react` (unchanged)
+
+User code would look like this:
+```tsx
+// server
+import { json } from "@remix-run/server";
+// adapter
+import { createRequestHandler, createFileSessionStorage } from "@remix-run/fly"
+// renderer
+import { useLoaderData } from "@remix-run/react";
+```
+
+### Internal packages
+
+- Runtime packages for authoring adapters (`@remix-run/{cloudflare,deno,node}`)
+- Package for authoring new runtimes (`@remix-run/server-runtime`), only including factory functions like:
+  - `createCookieFactory`
+  - `createSessionStorageFactory`
+  - `createCookieSessionStorageFactory`
+  - `createMemorySessionStorageFactory`
```

Comment:
I think `createRequestHandler` should still be here as well.

I agree with what @kiliman said in https://github.com/remix-run/remix/pull/3262#discussion_r877519747, that each adapter should have it's own `createRequestHandler`, but they all use the generic `createRequestHandler` (renamed to `createRemixRequestHandler` internally) under the hood, so we should keep that in `@remix-run/server-runtime` I think.

My reasoning in https://github.com/remix-run/remix/pull/3262#discussion_r877459283 about `createRequestHandler` only being specific in `@remix-run/deno` is actually wrong, as we combine both a runtime + adapter in 1 package.
If we would have created a separate `@remix-run/deno-deploy` adapter from the start, `@remix-run/deno` would have just re-exported `createRequestHandler` from `@remix-run/server-runtime`.